### ISSUE TITLE
Bug Report: App Crashes When Clicking Push Notification in Foreground in Android

### ISSUE THREAD

**Issue by abinashphulkonwar** [expo/expo#29878]
### Minimal reproducible example

https://github.com/abinashphulkonwar/expo-notification-crash-on-tap

### What platform(s) does this occur on?

Android

### Did you reproduce this issue in a development build?

Yes

### Summary

Our Expo-based React Native app crashes when a push notification is clicked while the app is in the foreground. The app works fine when the notification is clicked from the background or killed state.

**Error Message: from google play**
```
expo.modules.notifications.service.NotificationsService$Companion.createNotificationResponseBroadcastIntent
java.lang.IllegalArgumentException
```

**Error Message: from development client**
```our app just crashed. See the error below.
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.fam.famapp/expo.modules.notifications.service.NotificationForwarderActivity}: java.lang.IllegalArgumentException: notification and action should not be null
  android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3448)
  android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3595)
  android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
  android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
  android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
  android.app.ActivityThread$H.handleMessage(ActivityThread.java:2147)
  android.os.Handler.dispatchMessage(Handler.java:107)
  android.os.Looper.loop(Looper.java:237)
  android.app.ActivityThread.main(ActivityThread.java:7814)
  java.lang.reflect.Method.invoke(Native Method)
  com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
  com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1068)
Caused by java.lang.IllegalArgumentException: notification and action should not be null
  expo.modules.notifications.service.NotificationsService$Companion.createNotificationResponseBroadcastIntent(NotificationsService.kt:480)
  expo.modules.notifications.service.NotificationForwarderActivity.onCreate(NotificationForwarderActivity.kt:18)
  android.app.Activity.performCreate(Activity.java:7955)
  android.app.Activity.performCreate(Activity.java:7944)
  android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1307)
  android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3423)
  android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3595)
  android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
  android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
  android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
  android.app.ActivityThread$H.handleMessage(ActivityThread.java:2147)
  android.os.Handler.dispatchMessage(Handler.java:107)
  android.os.Looper.loop(Looper.java:237)
  android.app.ActivityThread.main(ActivityThread.java:7814)
  java.lang.reflect.Method.invoke(Native Method)
  com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
  com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1068)
```

**Steps to Reproduce:**

1. Open the app and ensure it is in the foreground.
2. Send a push notification to the device.
3. Click on the notification.
4. Observe the app crash.

**Expected Behavior:**
The app should handle the notification and navigate to the appropriate screen without crashing.

**Actual Behavior:**
The app crashes with an `java.lang.IllegalArgumentException` error.

**Code Snippet:**
Here's the relevant part of our notification handling code:

```
import React, {
  useState,
  useEffect,
  useReducer,
  createContext,
  useCallback,
  useRef,
} from "react";

import * as Notifications from "expo-notifications";
import { Platform } from "react-native";
import { useAppSelector } from "../../store/redux-toolkit/config";
import * as Device from "expo-device";
import Toast from "react-native-toast-message";
import Constants from "expo-constants";
import baseUrl, { chatBaseUrl } from "../../api/baseUrl";
import AssingStorage from "@react-native-async-storage/async-storage";
import { NotificationData } from "../../store/redux-toolkit/mainflow/notification";
import { useNavigation } from "@react-navigation/native";
import {
  NativeStackNavigationProp,
  NativeStackNavigatorProps,
} from "@react-navigation/native-stack/lib/typescript/src/types";
import { handleErrors } from "../../api/error-handler";
import { MessageScreenTabNamesEnum } from "../../screens/mainflow/messagesScreenFlow/messages";
import { messagedbInterface } from "../../store/redux-toolkit/mainflow/messages.type";
import { ParseError } from "../../error/request-error";
import { useDispatch } from "react-redux";
import { setMessagedb } from "../../store/redux-toolkit/mainflow/messages";
import { showToast } from "../../error/showTosta";

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: true,
    shouldSetBadge: false,
  }),
  // handleSuccess: async (id) => {
  //   console.log("success", id);
  // },
  // handleError: async (id) => {
  //   console.log("error", id);
  // },
});

const fetchMessagedb = async (id: string, token: string) => {
  const request = await fetch(`${chatBaseUrl}messagedb/${id}`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });
  if (request.ok) {
    return (await request.json()) as messagedbInterface;
  }
  const respons = await request.json();
  ParseError(respons, request.url);
  return null;
};

export const Notificationwrapper = ({ children }: React.PropsWithChildren) => {
  const refState = useRef({
    messagesdbIds: [] as messagedbInterface[],
  });
  const userRdeux = useAppSelector((state) => {
    refState.current.messagesdbIds = [
      ...state.MessagesRedux.chats.data,
      ...state.MessagesRedux.saveChatrooms.data,
    ];

    return state.personalAccount;
  });

  const reduxDispatch = useDispatch();

  const navigation = useNavigation<NativeStackNavigationProp<any>>();

  const pushNotificationListner = useCallback(
    async (notification: Notifications.Notification) => {
      try {
        let location = "";
        const event = notification.request.content.data as NotificationData;

        if (
          event.type === "message-chat" ||
          event.type === "message-saved-chatroom"
        ) {
          let messagedb = refState.current.messagesdbIds.find(
            (val) => val._id === event.resourceData.id
          );

          if (!messagedb) {
            messagedb = await fetchMessagedb(
              event.resourceData.id,
              userRdeux.token
            );
            if (!messagedb) {
              navigation.navigate("mainFlowScreen", {
                screen: "Message",
                params: {
                  initialTabName:
                    event.type === "message-chat"
                      ? MessageScreenTabNamesEnum["Chat"]
                      : MessageScreenTabNamesEnum["Saved Chatrooms"],
                },
              });
            }
            reduxDispatch(setMessagedb({ data: [messagedb] }));
          }
          if (messagedb.type == "group" || messagedb.type == "save-chatroom") {
            location = "ChatGroup";
          } else {
            location = "Chate";
          }

          navigation.navigate(location, messagedb);
        }
      } catch (error) {
        handleErrors(error);
      }
    },
    [userRdeux.token]
  );

  useEffect(() => {
    try {
      (async () => {
        if (!userRdeux.isLogIn) return;
        const isNotificationTokenSend = await AssingStorage.getItem(
          "isNotificationTokenGenerated"
        );
        console.log("device token: ", isNotificationTokenSend, " 🚀");
        //  if (isNotificationTokenSend) return;

        if (Platform.OS === "android") {
          await Notifications.setNotificationChannelAsync("default", {
            name: "default",
            importance: Notifications.AndroidImportance.MAX,
            vibrationPattern: [0, 250, 250, 250],
            lightColor: "#FF231F7C",
          });
        }
        if (Device.isDevice) {
          const { status: existingStatus } =
            await Notifications.getPermissionsAsync();
          let finalStatus = existingStatus;
          if (existingStatus !== "granted") {
            const { status } = await Notifications.requestPermissionsAsync();
            finalStatus = status;
          }
          if (finalStatus !== "granted") {
            Toast.show({
              type: "error",
              text1: "Please give the Permissions",
              text2: "Failed to get push token for push notification! 👋",
            });
            return;
          }
          const projectId =
            Constants?.expoConfig?.extra?.eas?.projectId ??
            Constants?.easConfig?.projectId;

          console.log("projectId: ", projectId);
          const token = (
            await Notifications.getExpoPushTokenAsync({
              projectId: projectId,
            })
          ).data;
          console.log("token: ", token);
          const fetchData = await fetch(`${baseUrl}/notification/token/set`, {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
              authorization: `Bearer ${userRdeux.token}`,
            },
            body: JSON.stringify({
              token: token,
             
            }),
          });
          const respons = await fetchData.json();
          console.log(respons);
          if (fetchData.ok) {
            console.log(token);
            await AssingStorage.setItem("isNotificationTokenGenerated", token);
          }
        } else {
          Toast.show({
            type: "error",
            text1: "Use physical device",
            text2: "Must use physical device for Push Notifications 👋",
          });
        }
      })();
    } catch (err) {
      console.log(err.message);
    }
  }, [userRdeux.isLogIn, userRdeux.token]);

  useEffect(() => {
    try {
     // const notificationListener =
      // Notifications.addNotificationReceivedListener(pushNotificationListner);

      const responseListener =
        Notifications.addNotificationResponseReceivedListener(
          async (response: Notifications.NotificationResponse) => {
            showToast("notification recived", 500);
            if (
              response.actionIdentifier ===
              Notifications.DEFAULT_ACTION_IDENTIFIER
            )
              await pushNotificationListner(response.notification);
          }
        );
      return () => {
     //   notificationListener.remove();
        responseListener.remove();
      };
    } catch (err) {
      console.log(err);
    }
  }, [pushNotificationListner]);

  return <>{children}</>;
};
```

### Environment

expo-env-info 1.2.0 environment info:
    System:
      OS: Windows 10 10.0.19045
    Binaries:
      Node: 20.9.0 - C:\Program Files\nodejs-20\node.EXE
      Yarn: 1.22.21 - ~\AppData\Roaming\npm\yarn.CMD
      npm: 10.1.0 - C:\Program Files\nodejs-20\npm.CMD
    IDEs:
      Android Studio: Version     2020.3.0.0 AI-203.7717.56.2031.7583922
    npmPackages:
      expo: ~50.0.19 => 50.0.19
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: 0.73.6 => 0.73.6
      react-native-web: ~0.19.6 => 0.19.7
    Expo Workflow: managed

### Expo Doctor Diagnostics

✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check native tooling versions
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check npm/ yarn versions
✔ Check for issues with metro config  
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check for legacy global CLI installed locally
✔ Check that native modules do not use incompatible support packages
✔ Check that native modules use compatible support package versions for installed Expo SDK
✔ Check that packages match versions required by installed Expo SDK

Didn't find any issues with the project!

**Comment by GrinZero** [comment]
This problem has been ongoing for several years

**Comment by GrinZero** [comment]
some issues like this:
- https://github.com/expo/expo/issues/24451
- https://github.com/expo/expo/issues/17531

**Comment by GrinZero** [comment]
Before PR is merged, anyone who sees this issue can try using `pnpm patch` or other patch methods for temporary support. Specific code reference: https://github.com/expo/expo/pull/30207

**Comment by this-is-bgm** [comment]
> Before PR is merged, anyone who sees this issue can try using `pnpm patch` or other patch methods for temporary support. Specific code reference: #30207

I am a rookie and I want to know how to patch him to work properly

**Comment by GrinZero** [comment]
> > Before PR is merged, anyone who sees this issue can try using `pnpm patch` or other patch methods for temporary support. Specific code reference: #30207
> 
> I am a rookie and I want to know how to patch him to work properly

read https://pnpm.io/cli/patch or https://github.com/ds300/patch-package

**Comment by ekremgunes** [comment]
I have the same similar issue..! When i get the notification on foreground app is crashing
 
```js
 useEffect(() => {
    const subscriptionReceived = Notifications.addNotificationReceivedListener(
      (notification) => {
        const content = notification.request.content;
        console.log(content)
      }
    );

    const subscriptionResponse =
      Notifications.addNotificationResponseReceivedListener((response) => {
        const eventId = response.notification.request.content.data.eventId;
        const screen = response.notification.request.content.data.screen;
        if (screen) {
          // navigation.navigate(screen, { eventId });
          setNavigateTo(screen, { eventId });
        }
      });

    return () => {
      subscriptionReceived.remove();
      subscriptionResponse.remove();
    };
  }, []);```

**Comment by nivashs** [comment]
Using Notifications.useLastNotificationResponse() for navigation instead of Notifications.addNotificationResponseReceivedListener fixed it for me.

**Comment by vonovak** [comment]
Hello and thanks for reporting,
this was resolved in #30207
Thank you 🙂
